chore: update docs (#264)
mineiros-io/terramate: docs/change-detection.md

# Change Detection

> When changing your infrastructure (made up of a set of stacks) it's common to
> make several changes to several stacks. But now that you have multiple terraform
> states (per stack), how to apply the changes only to the affected resources?

Reviewer: ```suggestion
When working with Infrastructure as Code it's considered to be a best practice to split up and organize your IaC into several smaller and isolated stacks. Typically, each stack comes with its own Terraform state which allows us to plan and apply each stack on its own. Thus, introducing change detection enables us to only plan and apply stacks that have changed and ignoring all others which can lead to significant faster execution time.
```

Reviewer: just a suggestion :

Author: It seems like it overlaps a little with why stacks indeed.

Reviewer: gotcha - feel free to adjust however you'd like to see this :)

Reviewer: I think we could use this as an introduction in the stacks part?

The terramate solves this by imposing a workflow:

> 1. The default branch (commonly main) has the production (applied) code.

Reviewer: ```suggestion
1. The default branch (commonly `main`) is considered to be the stable branch that represents the deployed state of our IaC.
```

> 2. Before planning and apply, the changes must be committed in a feature/bugfix
>   branch.

Reviewer: ```suggestion
2. Changes that should be planned and applied should be added through a feature or bugfix branch.
```

3. The IaC project must use [non
  fast-forwarded](https://git-scm.com/docs/git-merge#_fast_forward_merge) merge

>   commits. (the default in github and bitbucket).

Reviewer: ```suggestion
  commits. (the default in GitHub and Bitbucket).
```

## Why this workflow?

> By using the method above all commits (except first) in the default branch are

Author: Change Detection docs need to be reviewed/improved for sure, I just moved it from README to its own doc because it is an important concept.

> By using the method above all commits (except first) in the default branch are
> merge commits, then we have an easy way of detecting which stacks in the current
> feature branch differs from the main branch.

Reviewer: ```suggestion
Using the method above, all commits (except first) in the default branch
are merge commits, which provides us with an easy way of detecting
which stacks in the current feature branch differ from the main branch.
```

The technique is super simple and consists of finding the common ancestor

> between the current branch and the default branch (most commonly the commit that

Reviewer 2: I think this phrase needs some work because what it explains here is just change detection when in a feature branch. Maybe we can improve this in a separate PR but I think we should explain a bit the `baseRef` and how it has a role here. We can have a general explanation in terms of current revision (HEAD) and baseRef (as configured or defaultBranch or HEAD^ or --git-change-base).

Author: Lets do it in a separate PR

> The technique is super simple and consists of finding the common ancestor
> between the current branch and the default branch (most commonly the commit that
> current branch forked from) and then list the files that have changed since then
> and filter the ones that are part of terraform stacks.

Reviewer: ```suggestion
This technique is simple and consists of finding the common ancestor
between the current branch and the default branch (most commonly
the commit that the current branch forked from) and then list the files
that have changed since then and filter the ones that are part of Terraform stacks.
```

Author: Maybe Terramate stacks since Terraform don't actually have stacks ?
